tests/unit/optimization/StanFirstOrderSAGCostFunction_unittest.cc

```diff
@@ -0,0 +1,139 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#include <shogun/optimization/StanFirstOrderSAGCostFunction.h>
+#include "StanFirstOrderSAGCostFunction_unittest.h"
+#include <Eigen/Dense>
+#include <gtest/gtest.h>
+
+using namespace shogun;
+using Eigen::Matrix;
+using Eigen::Dynamic;
+using stan::math::var;
+using std::function;
+
+SGMatrix<float64_t> generateXSmall()
+{
+	SGMatrix<float64_t> X(2, 3);
+
+	X(0, 0) = 1;
+	X(0, 1) = 1;
+	X(0, 2) = 1;
+	X(1, 0) = 0;
+	X(1, 1) = 1;
+	X(1, 2) = 2;
+
+	return X;
+}
+
+StanVector generateParametersSmall()
+{
+	StanVector w(2, 1);
+	var m(1), c(0);
+	w(0, 0) = c;
+	w(1, 0) = m;
+
+	return w;
+}
+
+function<var(StanVector*, float64_t)>
+cost_for_ith_datapoint(SGMatrix<float64_t>& X, SGMatrix<float64_t>& y)
```

Comment:
Ok. So, I would like to understand this pattern based on declaring a function that returns a std::function, and the function's definition is a lambda. It feels like it is over-complicated.

Why not simply defining a function? Please explain.

Reply:
Just refactoring. I was using the same definition of a lambda twice in the file, so instead I just wrapped the lambda definition in a function: (cost_for_ith_datapoint) and used it in both places that were defining the lambdas so there is less copy paste.

Reply:
Just use a function that returns a 'var' (instead of std::function). For the definition of the function just use the logic in the body of the lambda.